File: tests/fixtures/deploy/python/server/probes.py
import hashlib
import json
import os
import re
import time
from urllib.parse import parse_qs, urlsplit

import six

MAX_SLEEP_MS = 30_000

MAX_BODY = 8 << 20

STREAM_CHUNKS = 5

STREAM_GAP = 0.2

STREAM_END = b"ocel-stream-end\n"

PROBE_HEADER = "x-ocel-probe"

CHECKSUM_FIELD = "x-ocel-sha256"

STATUS_PATH = re.compile(r"^/api/probes/status/(\d+)$")


def stream(req):
    req.start_chunks(
        {"content-type": "text/plain; charset=utf-8", "cache-control": "no-store, no-transform"}
    )
    for i in range(1, STREAM_CHUNKS):
        req.write_chunk(f"ocel-stream-{i}\n".encode())
        time.sleep(STREAM_GAP)
    req.write_chunk(STREAM_END)
    req.end_chunks()


def status(req):
    code = int(STATUS_PATH.match(path_of(req)).group(1))
    if code == 204:
        req.write_status(204)
        return
    extra = {"location": "/api/probes/status/204"} if 300 <= code < 400 else None
    req.write_bytes(code, "application/json", json.dumps({"status": code}).encode(), extra)


def echo(req):
    read = req.read_body()
    if read is None:
        refuse(req)
        return
    query = {name: values[0] for name, values in parse_qs(urlsplit(req.path).query).items()}
    req.write_json(
        200,
        {
            "method": req.command,
            "path": path_of(req),
            "query": query,
            "header": req.headers.get(PROBE_HEADER),
            "body": echoed(req, read),
        },
    )


def echoed(req, read):
    if not read:
        return None
    if "application/json" in (req.headers.get("content-type") or ""):
        try:
            return json.loads(read)
        except ValueError:
            pass
    return read.decode("utf-8", "replace")


def take_large(req):
    read = req.read_body()
    if read is None:
        refuse(req)
        return
    req.write_json(200, {"bytes": len(read), "sha256": hashlib.sha256(read).hexdigest()})


def send_large(req):
    wanted = number_of(req, "bytes")
    if wanted is None or wanted < 0 or wanted > MAX_BODY:
        req.write_json(400, {"error": f"bytes must be an integer between 0 and {MAX_BODY}"})
        return
    body = os.urandom(wanted)
    req.write_bytes(
        200,
        "application/octet-stream",
        body,
        {CHECKSUM_FIELD: hashlib.sha256(body).hexdigest()},
    )


def sleep(req):
    ms = number_of(req, "ms")
    if ms is None or ms < 0 or ms > MAX_SLEEP_MS:
        req.write_json(400, {"error": f"ms must be an integer between 0 and {MAX_SLEEP_MS}"})
        return
    time.sleep(ms / 1000)
    req.write_json(200, {"slept": ms})


def vendored(req):
    req.write_json(
        200,
        {
            "dependency": "six",
            "version": six.__version__,
            "answer": len(list(six.moves.range(2))),
        },
    )


def refuse(req):
    req.write_bytes(
        400,
        "application/json",
        json.dumps({"error": "bad request"}).encode(),
        {"connection": "close"},
    )


def path_of(req):
    return req.path.split("?", 1)[0]


def number_of(req, name):
    values = parse_qs(urlsplit(req.path).query).get(name)
    if not values:
        return 0
    try:
        return int(values[0])
    except ValueError:
        return None


def at(method, path):
    return lambda seen, asked: seen == method and asked == path


def under(path):
    return lambda _, asked: asked == path or asked.startswith(path + "/")


PROBES = [
    (at("GET", "/api/probes/stream"), stream),
    (lambda seen, asked: seen == "GET" and STATUS_PATH.match(asked) is not None, status),
    (under("/api/probes/echo"), echo),
    (at("POST", "/api/probes/large"), take_large),
    (at("GET", "/api/probes/large"), send_large),
    (at("GET", "/api/probes/sleep"), sleep),
    (at("GET", "/api/probes/vendored"), vendored),
]

File: tests/fixtures/deploy/go/server/probes.go
package main

import (
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strconv"
	"strings"
	"time"
)

const (
	maxSleep      = 30 * time.Second
	maxBody       = 8 << 20
	streamChunks  = 5
	streamGap     = 200 * time.Millisecond
	streamEnd     = "ocel-stream-end"
	probeHeader   = "x-ocel-probe"
	checksumField = "x-ocel-sha256"
)

func probes(mux *http.ServeMux) {
	mux.HandleFunc("GET /api/probes/stream", stream)
	mux.HandleFunc("GET /api/probes/status/{code}", status)
	mux.HandleFunc("/api/probes/echo", echo)
	mux.HandleFunc("/api/probes/echo/", echo)
	mux.HandleFunc("POST /api/probes/large", takeLarge)
	mux.HandleFunc("GET /api/probes/large", sendLarge)
	mux.HandleFunc("GET /api/probes/sleep", sleep)
}

func stream(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("content-type", "text/plain; charset=utf-8")
	w.Header().Set("cache-control", "no-store, no-transform")
	w.WriteHeader(http.StatusOK)

	flush := http.NewResponseController(w)
	for i := 1; i < streamChunks; i++ {
		fmt.Fprintf(w, "ocel-stream-%d\n", i)
		if err := flush.Flush(); err != nil {
			return
		}
		select {
		case <-time.After(streamGap):
		case <-r.Context().Done():
			return
		}
	}
	fmt.Fprintln(w, streamEnd)
}

func status(w http.ResponseWriter, r *http.Request) {
	code, err := strconv.Atoi(r.PathValue("code"))
	if err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]any{"error": "code must be a status"})
		return
	}
	if code >= 300 && code < 400 {
		w.Header().Set("location", "/api/probes/status/204")
	}
	if code == http.StatusNoContent {
		w.WriteHeader(code)
		return
	}
	writeJSON(w, code, map[string]any{"status": code})
}

func echo(w http.ResponseWriter, r *http.Request) {
	read, err := io.ReadAll(http.MaxBytesReader(w, r.Body, maxBody))
	if err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]any{"error": "bad request"})
		return
	}

	query := map[string]string{}
	for key, values := range r.URL.Query() {
		query[key] = values[0]
	}

	var header any
	if named := r.Header.Get(probeHeader); named != "" {
		header = named
	}

	writeJSON(w, http.StatusOK, map[string]any{
		"method": r.Method,
		"path":   r.URL.Path,
		"query":  query,
		"header": header,
		"body":   echoedBody(r, read),
	})
}

func echoedBody(r *http.Request, read []byte) any {
	if len(read) == 0 {
		return nil
	}
	if strings.Contains(r.Header.Get("content-type"), "application/json") {
		var decoded any
		if json.Unmarshal(read, &decoded) == nil {
			return decoded
		}
	}
	return string(read)
}

func takeLarge(w http.ResponseWriter, r *http.Request) {
	read, err := io.ReadAll(http.MaxBytesReader(w, r.Body, maxBody))
	if err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]any{"error": "bad request"})
		return
	}
	sum := sha256.Sum256(read)
	writeJSON(w, http.StatusOK, map[string]any{
		"bytes":  len(read),
		"sha256": hex.EncodeToString(sum[:]),
	})
}

func sendLarge(w http.ResponseWriter, r *http.Request) {
	bytes, err := strconv.Atoi(numberOr(r, "bytes", "0"))
	if err != nil || bytes < 0 || bytes > maxBody {
		writeJSON(w, http.StatusBadRequest, map[string]any{
			"error": fmt.Sprintf("bytes must be an integer between 0 and %d", maxBody),
		})
		return
	}
	body := make([]byte, bytes)
	if _, err := rand.Read(body); err != nil {
		writeJSON(w, http.StatusInternalServerError, map[string]any{"error": "internal error"})
		return
	}
	sum := sha256.Sum256(body)
	w.Header().Set("content-type", "application/octet-stream")
	w.Header().Set(checksumField, hex.EncodeToString(sum[:]))
	w.Header().Set("content-length", strconv.Itoa(len(body)))
	w.WriteHeader(http.StatusOK)
	_, _ = w.Write(body)
}

func sleep(w http.ResponseWriter, r *http.Request) {
	ms, err := strconv.Atoi(numberOr(r, "ms", "0"))
	if err != nil || ms < 0 || time.Duration(ms)*time.Millisecond > maxSleep {
		writeJSON(w, http.StatusBadRequest, map[string]any{
			"error": fmt.Sprintf("ms must be an integer between 0 and %d", maxSleep.Milliseconds()),
		})
		return
	}
	select {
	case <-time.After(time.Duration(ms) * time.Millisecond):
	case <-r.Context().Done():
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"slept": ms})
}

func numberOr(r *http.Request, name, fallback string) string {
	if named := r.URL.Query().Get(name); named != "" {
		return named
	}
	return fallback
}
